### PR TITLE
Add constraints

### PR DESCRIPTION
**Pull request by wndhydrnt**
This PR adds support for `constraints` as described in #256.

The format to define constraints of a job is the same as in Marathon.
Note that currently only the `CLUSTER` constraint has been implemented.

### CODE REVIEW

**Comment by elingg** [comment]
Nice @wndhydrnt, will take a look ASAP!

**Comment by elingg** [comment]
Looking good. One overall comment to begin wtih - could we name this constraints or job constraints? Cluster constraints make it seem like it's for all jobs in the cluster.

**Comment by wndhydrnt** [comment]
The implementation is similar to how [Marathon defines constraints](https://mesosphere.github.io/marathon/docs/constraints.html).
The `ClusterConstraint` that has the operator `CLUSTER` is just one implementation of the basic trait `Constraint`. Looking at the constraints supported by Marathon, the `LIKE` constraint could be useful and can be added later.
From that point of view, I would not change the naming but the docs may need to provide more details.

However, it all depends on whether there will be other constraints that need to be implemented. If you think that this is not very likely, I would suggest the following:

- Remove the `CLUSTER` operator as it is kind of pointless in this case
- Use a hash as the value of `constraints` in the configuration of a job, like
```json
...
"constraints": {"rack": "rack-1", "profile": "high-load"},
...
```
- Implement a `Constraint` class instead of a trait

What do you think @elingg ?

**Comment by elingg** [comment]
Ah, in Marathon cluster constraints are for cluster wide constraints, which I think we should also have for Chronos eventually: 

```CLUSTER allows you to run all of your app's tasks on slaves that share a certain attribute. This is useful for example if you have apps with special hardware needs, or if you want to run them on the same rack for low latency.```

It seems the only implementation you have here is cluster wide constraints, but these are implemented at the job level. I see this in your code:

```
  offerResources.toIterator.find { ors =>
+                  ors._2.canSatisfy(neededResources) && checkConstraints(ors._1.getAttributesList.asScala, job.constraints)
```




Cluster wide constraints should be not be set and checked at the job level.

**Comment by wndhydrnt** [comment]
I think I got your point. The difference is that a Marathon app can have several instances (= tasks) running at the same time where with Chronos there is only one task of a job running at the same time.

Here is an example, just to make clear what I want to achieve.
Given five Mesos slaves with the following attributes:

Slave | Attributes
------- | -------------
s0 | rack:a
s1 | rack:a
s2 | rack:b
s3 | rack:b
s4 | rack:c

And a job with `[["rack", "CLUSTER", "b"]]` set as the value of the `constraints` key.
Chronos should make sure that tasks of this job always run on `s2` or `s3` and never on `s0`, `s1` or `s4`.

---

I also thought of a different name for the operator instead of `CLUSTER`. Maybe something like `EQUAL` or `EQUALS` better reflects what the constraint is doing.

**Comment by elingg** [comment]
Yes, @wndhydrnt, this makes sense and I agree. I think the EQUALS operator would be appropriate here.

**Comment by wndhydrnt** [comment]
:+1: I'll change the naming and update the PR

**Comment by elingg** [comment]
Great, nice work! I love that we are putting the first round of constraints into Chronos.

**Comment by wndhydrnt** [comment]
Done. Please take a look at it again @elingg.

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/Jobs.scala [hunk 1]
```diff
@@ -110,7 +113,8 @@ case class ScheduleBasedJob(
                              @JsonProperty override val shell: Boolean = true,
                              @JsonProperty override val arguments: Seq[String] = List(),
                              @JsonProperty override val softError: Boolean = false,
-                             override val dataProcessingJobType: Boolean = false)
```
Comment (elingg) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/Jobs.scala]: not your code, but could we add @JsonProperty to dataProcessingJobType while you are at it. Not sure why this was left off.

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/constraints/EqualsConstraint.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+case class EqualsConstraint(attribute: String, value: String) extends Constraint {
+
+  def matches(attributes: Seq[Protos.Attribute]): Boolean = {
+    attributes.foreach { a =>
+      if (a.getName == attribute && a.getText.getValue == value) {
+        return true
+      }
+    }
+
```
Comment (elingg) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/constraints/EqualsConstraint.scala]: Why do we have a space here?

File: src/main/scala/org/apache/mesos/chronos/scheduler/mesos/MesosJobFramework.scala [hunk 3]
```diff
@@ -99,6 +100,16 @@ import scala.concurrent.ExecutionContext.Implicits.global
   def generateLaunchableTasks(offerResources: mutable.HashMap[Offer, Resources]): mutable.Buffer[(String, BaseJob, Offer)] = {
     val tasks = mutable.Buffer[(String, BaseJob, Offer)]()
 
+    def checkConstraints(attributes: Seq[Protos.Attribute], constraints: Seq[Constraint]): Boolean = {
+      constraints.foreach { c =>
+        if (!c.matches(attributes)) {
+          return false
+        }
+      }
+
```
Comment (elingg) [inline src/main/scala/org/apache/mesos/chronos/scheduler/mesos/MesosJobFramework.scala]: Why is there an extra space here

**Comment by elingg** [comment]
Just a couple of nits. LGTM

**Comment by wndhydrnt** [comment]
Fixed the issues.

**Comment by elingg** [comment]
:+1:  Any objections, @brndnmtthws?

**Comment by brndnmtthws** [comment]
Looks good to me. One day I'd like to see Marathon and Chronos sharing the same constraints code.

**Comment by kangman** [comment]
Hi when will this be available in a stable release?

**Comment by elingg** [comment]
Since we've gotten a number of changes and bug fixes in, we will be releasing a new version of Chronos (2.3.4) in the next week or two.

**Comment by kangman** [comment]
Thanks @elingg  just fyi the documentation gave me some false hope that we already had this feature.  Here is the [entry](https://mesos.github.io/chronos/docs/api.html#sample-job).

**Comment by elingg** [comment]
You're in luck. The release should be out sooner than expected. We're looking at this week.

**Comment by kangman** [comment]
Thanks @elingg !!  Made my day~

**Comment by elingg** [comment]
:+1:

**Pull request merged**